### PR TITLE
[WIP] Feature: review code output options

### PR DESCRIPTION
**Pull request by sapientcoffee**
The option to output `devai review code` in the existing markdown (remain default) format along with json and in a table. This change adds `-o` with the options `markdown`, `json`, `table`.

To achieve this the prompts have also been externalised to their own individual files (including ones that define the required output for json and markdown).

The default is markdown to allow for backwards compatibility.

$ devai review code -c ../sample-app/src/main/java

> ```
> Response from Model: ## Code Review: Bank of Anthos - BalanceReader
> 
> **Overall:** The BalanceReader code demonstrates a good understanding of Spring Boot, caching, and security principles. However, there are some areas for potential improvement in terms of error handling, thread management, and code style.
> 
> **Transaction.java**
> 
> * **No Issues Found:** This class is well-structured and straightforward. 
> 
> **LedgerReader.java**
> 
> * **Thread Management (Best Practices):** The `backgroundThread` is managed within the `LedgerReader` class. While this works, it might be more robust to delegate thread management to a dedicated `ExecutorService` for better control over thread lifecycle and potential thread pool management in the future.
>    * **Suggestion:** Consider using `ScheduledExecutorService` to manage the polling task, providing a more structured way to schedule and control the thread.
> 
> * **Error Handling (Robustness):** The `pollTransactions` method silently catches `ResourceAccessException` and `DataAccessResourceFailureException`. While the intention might be to retry, consider logging the exceptions with more details (e.g., stack trace) or adding retry logic with exponential backoff for transient errors. 
> 
> * **Magic Number (Code Style):** Using `-1` as `STARTING_TRANSACTION_ID` could be made more expressive. 
>    * **Suggestion:** Define a constant like `private static final long NO_TRANSACTIONS = -1L;` to improve readability.
> 
> **BalanceCache.java**
> 
> * **No Major Issues Found:** The use of Guava's `LoadingCache` is appropriate, and the configuration appears reasonable. 
> 
> **JWTVerifierGenerator.java**
> 
> * **Key Loading (Security/Best Practices):** The public key is loaded from the file system on each request. In a production environment, consider caching the loaded key for performance reasons.
>     * **Suggestion:** Load the public key once during application startup and store it in memory for subsequent use.  You might use a library like `jasypt-spring-boot` for secure configuration loading if needed. 
> 
> **BalanceReaderApplication.java**
> 
> * **Environment Variable Check (Robustness):** The environment variable check during startup is good practice. However, consider using a more structured approach for configuration validation, perhaps using Spring Boot's `@ConfigurationProperties` and `@Validated` annotations to leverage Spring's validation framework.
> 
> **BalanceReaderController.java**
> 
> * **Bearer Token Handling (Code Style):**  The manual parsing of the `bearerToken` could be simplified. Spring Security provides utilities for handling bearer tokens more cleanly. 
>    * **Suggestion:** If you're open to using Spring Security, leverage its `BearerTokenAuthenticationFilter` or related components for token extraction and validation.
> 
> * **Error Responses (Best Practices):** Returning string literals like `"not authorized"` and `"cache error"` might not be ideal for all clients. Consider returning more structured error responses, possibly with error codes, to make it easier for clients to handle different error scenarios.
> 
> **General Observations:**
> 
> * **Logging (Maintainability):** The logging is generally good. However, consider adding more detailed log messages in critical areas like exception handling to aid in debugging.
> * **Testing (Best Practices):** While not directly part of the review, ensure you have sufficient unit and integration tests covering the logic, especially error handling and edge cases. 
> 
> Let me know if you'd like a more in-depth analysis of a specific aspect or have any particular concerns. 
>```

$ devai review code -c ../sample-app/src/main/java -o json

> ```
> [
>     {
>         "class": "BalanceReaderController",
>         "method": "getBalance",
>         "issue": "Potential Cache Miss Handling",
>         "severity": "Medium",
>         "description": "The `getBalance` method retrieves the balance from the cache using `cache.get(accountId)`. While the `LoadingCache` should ideally handle cache misses, it's worth considering explicitly handling cases where `cache.get()` might throw an exception (e.g., due to a backend data store issue). This could involve logging the error and returning an appropriate error response to the client, preventing potential application crashes."
>     },
>     {
>         "class": "LedgerReader",
>         "method": "startWithCallback",
>         "issue": "Infinite Loop in Background Thread",
>         "severity": "Critical",
>         "description": "The background thread in `startWithCallback` has a potential issue. If an exception occurs within the `try` block (e.g., `InterruptedException`, `ResourceAccessException`, etc.), the `alive` flag might not be set to `false`. This could result in an infinite loop where the thread keeps running but doesn't process transactions, potentially leading to resource exhaustion. It's recommended to ensure that `alive` is set to `false` in all exception scenarios to gracefully stop the background thread."
>     },
>     {
>         "class": "LedgerReader",
>         "method": "startWithCallback",
>         "issue": "Resource Management: Thread Interruption",
>         "severity": "Medium",
>         "description": "In the `startWithCallback` method, the background thread doesn't handle thread interruption cleanly. If the thread is interrupted while sleeping (e.g., during application shutdown), the `InterruptedException` is caught and logged, but the thread continues running. It's best practice to honor thread interruption and exit gracefully. Consider re-throwing the `InterruptedException` or setting a flag to indicate that the thread should terminate."
>     },
>     {
>         "class": "BalanceReaderController",
>         "method": "getBalance",
>         "issue": "Exception Handling and Logging",
>         "severity": "Low",
>         "description": "In the `getBalance` method, the catch blocks for `JWTVerificationException` and `ExecutionException/UncheckedExecutionException` log the error but return generic error messages to the client. While this is acceptable for security reasons (not revealing too much information about the error), consider providing more informative logging, including the exception message or stack trace, to aid in debugging potential issues."
>     },
>     {
>         "class": "LedgerReader",
>         "method": "pollTransactions",
>         "issue": "Potential Logging Overhead",
>         "severity": "Low",
>         "description": "The `pollTransactions` method logs \"Polling Transactions\" in debug mode. If this method is called frequently, it might lead to excessive logging, especially if the debug level is enabled in production. Consider logging at a less verbose level (e.g., trace or info) or logging only when necessary, such as when a new transaction is found."
>     }
> ]
> ```

$ devai review code -c ../sample-app/src/main/java -o table

> ```
> ┏━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┓
> ┃ class                   ┃ method           ┃ question                             ┃ answer                                                                                                                                    ┃ severity ┃
> ┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━┩
> │ General                 │ N/A              │ Overall Impression                   │ The code is well-structured and demonstrates good use of caching and background processing.  However, there are opportunities to improve  │ Medium   │
> │                         │                  │                                      │ error handling and logging.                                                                                                               │          │
> │ LedgerReader            │ run              │ Handling Remote Database Out of Sync │ When the remote database is out of sync, the application stops processing transactions. Would it be more resilient to implement a retry   │ High     │
> │                         │                  │                                      │ mechanism or a more sophisticated synchronization strategy instead of completely halting the processing?                                  │          │
> │ BalanceReaderController │ getBalance       │ JWT Verification Exception Handling  │ The code catches JWTVerificationException but returns a generic 'not authorized' message for different verification failures. Could we    │ Medium   │
> │                         │                  │                                      │ provide more specific error messages (e.g., expired token, invalid signature) to aid in debugging and client-side error handling?         │          │
> │ BalanceReaderController │ getBalance       │ Cache Error Handling                 │ Similar to the JWT verification, the cache error handling returns a generic message.  It might be beneficial to log the specific          │ Medium   │
> │                         │                  │                                      │ exception and provide a more informative error message or potentially attempt a fallback mechanism.                                       │          │
> │ LedgerReader            │ pollTransactions │ Logging Transaction Details          │ Currently, the logging in 'pollTransactions' only indicates the polling action.  Would it be helpful to log the transaction details being │ Low      │
> │                         │                  │                                      │ processed for debugging and auditing purposes?                                                                                            │          │
> │ BalanceCache            │ load             │ Cache Loading Logging                │ The log message 'Cache loaded from db' in the 'load' method might benefit from including the accountId for which the cache is being       │ Low      │
> │                         │                  │                                      │ loaded.  This could help in tracing cache behavior during debugging.                                                                      │          │
> └─────────────────────────┴──────────────────┴──────────────────────────────────────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴──────────┘
> ```

### CODE REVIEW

File: devai-cli/src/requirements.txt [hunk 1]
```diff
@@ -7,6 +7,9 @@ langchain-community==0.0.38
 langchain_google_vertexai==0.0.5
 atlassian-python-api==3.41.10
 chromadb==0.5.0
+json
+rich
+json-repair
```
Comment (gitrey) [inline devai-cli/src/requirements.txt:12]: @sapientcoffee could you please add version numbers for new dependencies?

Comment (gitrey) [reply devai-cli/src/requirements.txt:12]: Also got this error when I tried to install.

```
GoogleCloudPlatform/genai-for-developers/devai-cli  feature-review-code-outputs ✗                                                                                         1d0h ◒  ⍉
▶ pip install -r src/requirements.txt
Requirement already satisfied: click==8.1.7 in /Users/andreyshakirov/work/github/langsmith-tracing/venv/lib/python3.11/site-packages (from -r src/requirements.txt (line 1)) (8.1.7)
Requirement already satisfied: google-cloud-aiplatform==1.48.0 in /Users/andreyshakirov/work/github/langsmith-tracing/venv/lib/python3.11/site-packages (from -r src/requirements.txt (line 2)) (1.48.0)
Requirement already satisfied: jira==3.6.0 in /Users/andreyshakirov/work/github/langsmith-tracing/venv/lib/python3.11/site-packages (from -r src/requirements.txt (line 3)) (3.6.0)
Requirement already satisfied: python-gitlab==4.4.0 in /Users/andreyshakirov/work/github/langsmith-tracing/venv/lib/python3.11/site-packages (from -r src/requirements.txt (line 4)) (4.4.0)
Requirement already satisfied: langchain==0.1.8 in /Users/andreyshakirov/work/github/langsmith-tracing/venv/lib/python3.11/site-packages (from -r src/requirements.txt (line 5)) (0.1.8)
Requirement already satisfied: langchain-community==0.0.38 in /Users/andreyshakirov/work/github/langsmith-tracing/venv/lib/python3.11/site-packages (from -r src/requirements.txt (line 6)) (0.0.38)
Requirement already satisfied: langchain_google_vertexai==0.0.5 in /Users/andreyshakirov/work/github/langsmith-tracing/venv/lib/python3.11/site-packages (from -r src/requirements.txt (line 7)) (0.0.5)
Requirement already satisfied: atlassian-python-api==3.41.10 in /Users/andreyshakirov/work/github/langsmith-tracing/venv/lib/python3.11/site-packages (from -r src/requirements.txt (line 8)) (3.41.10)
Requirement already satisfied: chromadb==0.5.0 in /Users/andreyshakirov/work/github/langsmith-tracing/venv/lib/python3.11/site-packages (from -r src/requirements.txt (line 9)) (0.5.0)
ERROR: Could not find a version that satisfies the requirement json (from versions: none)
ERROR: No matching distribution found for json

[notice] A new release of pip is available: 23.2.1 -> 24.0
[notice] To update, run: pip install --upgrade pip
(venv) 
GoogleCloudPlatform/genai-for-developers/devai-cli  feature-review-code-outputs ✗   
```

**Comment by sapientcoffee** [comment]
replaced with #35
